xbatcher/generators.py

```diff
@@ -167,22 +178,20 @@ def _gen_batches(self) -> dict:
         # going the eager route for now is allowing me to fill out the loader api
         # but it is likely to perform poorly.
         batches = []
-        for ds_batch_selector in self._iterate_batch_dims(self.ds):
+        for ds_batch_selector in self._iterate_batch_dims():
             ds_batch = self.ds.isel(**ds_batch_selector)
             if self.preload_batch:
                 ds_batch.load()
-
-            input_generator = self._iterate_input_dims(ds_batch)
-
+            input_generator = self._iterate_input_dims()
             if self.concat_input_dims:
                 batches.append(list(input_generator))
             else:
                 batches += list(input_generator)
 
-        return dict(zip(range(len(batches)), batches))
+        return dict(enumerate(batches))
 
-    def _iterate_batch_dims(self, ds):
-        return _iterate_through_dataset(ds, self.batch_dims)
+    def _iterate_batch_dims(self) -> Any:
```

Comment:
```suggestion
    def _iterate_batch_dims(self) -> Generator[Dict[str, int], None, None]:
```

I didn't test this but I think it is basically right. If it works, you can propagate this to the `_iterate_through_dataset`  function as well.

Reply:
Alternatively, you can use the `Iterator` type.

Reply:
Lots of errors using `def _iterate_batch_dims(self) -> Iterator[Dict[Hashable, slice]]:`.
```
xbatcher/generators.py:182: error: Keywords must be strings
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "Optional[Mapping[Any, Any]]"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "bool"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "Literal['raise', 'warn', 'ignore']"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "Optional[Mapping[Any, Any]]"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "bool"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "Literal['raise', 'warn', 'ignore']"
xbatcher/generators.py:189: error: Argument 1 to "list" has incompatible type "Iterator[Dict[Hashable, slice]]"; expected "Iterable[List[Dict[Hashable, slice]]]"
Found 8 errors in 1 file (checked 20 source files)
```
I have a separate refactor almost finished that would fix https://github.com/xarray-contrib/xbatcher/issues/131 and make this function obsolete. Do view it as necessary to type all functions before merging in the existing type hints? My preference would be to not spend more time on this.